distribution/push_v2.go

```diff
@@ -425,26 +422,34 @@ func (pd *v2PushDescriptor) uploadUsingSession(
 	diffID layer.DiffID,
 	layerUpload distribution.BlobWriter,
 ) (distribution.Descriptor, error) {
-	arch, err := pd.layer.TarStream()
-	if err != nil {
-		return distribution.Descriptor{}, xfer.DoNotRetry{Err: err}
-	}
-
-	// don't care if this fails; best effort
-	size, _ := pd.layer.DiffSize()
-
-	reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, arch), progressOutput, size, pd.ID(), "Pushing")
-	compressedReader, compressionDone := compress(reader)
-	defer func() {
+	var (
+		reader io.ReadCloser
+	)
+
+	contentReader, err := pd.layer.Open()
+	size := pd.layer.Size()
+
+	reader = progress.NewProgressReader(ioutils.NewCancelReadCloser(ctx, contentReader), progressOutput, size, pd.ID(), "Pushing")
+
+	switch m := pd.layer.MediaType(); m {
+	case "application/vnd.docker.image.rootfs.diff.tar":
```

Comment:
The content types are not very consistent. Really we have plugin images and container images, but for the name roots we have 

`application/vnd.docker.container.image` - for container image configs
`application/vnd.docker.plugin` - for plugin configs
`application/vnd.docker.image` - for image layers

The plugin image layers are the same as container image layers though and should be able to share a backend.